contributors/design-proposals/customresources-validation.md

```diff
@@ -0,0 +1,493 @@
+2. Add a validation step to the CREATE and UPDATE REST handlers of the apiextensions-apiserver.
+ 
+As an optional follow-up, we can implement the OpenAPI part and with that enable client-side validation:
+
+3. Export the JSON-Schema via a dynamically served OpenAPI spec.
+
+## Appendix
+
+### Expressiveness of JSON-Schema
+
+The following example properties cannot be expressed using JSON-Schema:
+1. “In a PodSpec, for each `spec.Containers[*].volumeMounts[*].Name` there must be a `spec.Volumes[*].Name`”
+2. “The volume names in `PodSpec.Volumes` are unique” (`uniqueItems` only compares the complete objects, it cannot compare by key)
+ 
+Different versions within one CRD with a custom version field (i.e. not the one in apiVersion) **can** be expressed:
+
+```json
+{
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "title": "child_schema",
+    "type": "object",
+    "anyOf": [
+        {
+            "properties": {
+                "version": {
+                    "type": "string",
+                    "pattern": "^a$"
+                },
+                "spec": {
+                    "type": "object",
+                    "properties": {
+                        "foo": {}
+                    },
+                    "additionalProperties": false,
+                },
+            }
+        },
+        {
+            "properties": {
+                "version": {
+                    "type": "string",
+                    "pattern": "^b$"
+                },
+                "spec": {
+                    "type": "object",
+                    "properties": {
+                        "bar": {}
+                    },
+                    "additionalProperties": false,
+                },
+            }
+        }
+    ],
+}
+```
+
+This validates: 
+* `{"version": "a",  "spec": {"foo": 42}}`
+* `{"version": "b",  "spec": {"bar": 42}}`
+
+but not:
+* `{"version": "a",  "spec": {"bar": 42}}`.
+ 
+Note: this is a workaround while we do not support multiple versions and conversion for custom resources.
+
+### JSON-Schema Validation Runtime Complexity
+
+Following “JSON: data model, query languages and schema specification<sup id="f4">[4](#footnote4)</sup>” and “Formal Specification, Expressiveness and Complexity analysis for JSON Schema<sup id="f5">[5](#footnote5)</sup>”, JSON-Schema validation
+* without the uniqueItems operator and 
+* without recursion for the $ref operator
+has linear runtime in the size of the JSON input and the size of the schema (Th. 1 plus Prop. 7).
+ 
+If we allow uniqueItems, the runtime complexity becomes quadratic in the size of the JSON input. Hence, we might want to consider forbidding the uniqueItems operator in order to avoid DDoS attacks, at least if the schema definitions of CRDs cannot be trusted.
+
+The CRD JSON-Schema will be validated to have neither recursion, nor `uniqueItems=true` being set.
+
+### Alternatives
```

Comment:
Schema validation performed by kubectl is driven by Swagger/OpenAPI models. An equivalent capability needs to move to the server (https://github.com/kubernetes/kubernetes/issues/5889) as part of the general effort to move functionality to the server in order to improve extensibility and to simplify clients (https://github.com/kubernetes/kubernetes/issues/12143).

I would like to see us move towards declarative validation generally for the API, as much as possible:
https://github.com/kubernetes/kubernetes/issues/25460

@mbohlool may have already started on a proposal for that.

Also, ideally, I'd like the same spec to be usable to serve an OpenAPI spec for the CRD APIs, which would mean it would have to be an OpenAPI-compatible flavor of JSON schema. I really don't want multiple different schema languages in the system.

As for a hook-based approach, (potentially multiple) special-purpose hooks per resource would be harder to understand, as well as being more fragile, slower, etc. I'd prefer to find a way to use admission-control extension for that use case, so as not to add yet another hook mechanism:

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/admission_control_extension.md